Add a ProcMount option to the SecurityContext & AllowedProcMountTypes to PodSecurityPolicy

jessfraz opened the pull request:
So there is a bit of a chicken and egg problem here in that the CRI runtimes will need to implement this for there to be any sort of e2e testing.

**What this PR does / why we need it**: This PR implements design proposal https://github.com/kubernetes/community/pull/1934. This adds a ProcMount option to the SecurityContext and AllowedProcMountTypes to PodSecurityPolicy

Relies on https://github.com/google/cadvisor/pull/1967

**Release note**:
<!--  Write your release note:
1. Enter your extended release note in the below block. If the PR requires additional action from users switching to the new release, include the string "action required".
2. If no release note is required, just write "NONE".
-->
```release-note
ProcMount added to SecurityContext and AllowedProcMounts added to PodSecurityPolicy to allow paths in the container's /proc to not be masked.
```

cc @Random-Liu @mrunalp

jessfraz committed:
pkg/kubelet/apis/cri/runtime: add masked_paths and readonly_paths

generate runtime protobufs

Signed-off-by: Jess Frazelle <acidburn@microsoft.com>

jessfraz committed:
ProcMount: add api options and feature gate

Signed-off-by: Jess Frazelle <acidburn@microsoft.com>

jessfraz committed:
ProcMount: update staging types

Signed-off-by: Jess Frazelle <acidburn@microsoft.com>

dims commented on pkg/security/podsecuritypolicy/provider.go:
```diff
@@ -271,6 +271,21 @@ func (s *simpleProvider) ValidateContainerSecurityContext(pod *api.Pod, containe
 
+	procMount := sc.ProcMount()
+	allowedProcMounts := s.psp.Spec.AllowedProcMountTypes
+	if allowedProcMounts == nil {
```
Check for empty as well? (in addition to nil)

annismckenzie replied:
So a length check against 0 would catch both.

dims commented:
@jessfraz we need a Feature for this right? (entry in kube_features.go)

jessfraz commented:
Oh thanks for reminding me I'll do that

AkihiroSuda commented on pkg/kubelet/apis/cri/runtime/v1alpha2/api.proto:
```diff
@@ -586,6 +586,12 @@ message LinuxContainerSecurityContext {
     bool no_new_privs = 11;
+    // masked_paths is a slice of paths that should be masked by the container
+    // runtime, this can be passed directly to the OCI spec.
+    repeated string masked_paths = 13;
```
nit: 13 -> 12?

dims replied:
y, i scratched my head for a few mins and realized that

runcom replied:
is there any guidance for runtimes on how to handle these two fields? https://github.com/containerd/cri/commit/3e4cec8739bbb0db0764f6db80a25a3f75411c0b#diff-c656bacd6cb0fe9a7f64814b01256decR358 cri-containerd seems to handle it that way but I believe there must be guidance in the api itself, @Random-Liu @jessfraz

dims commented:
/unassign
/uncc

jessfraz committed:
update jsonlog path for updated vendor

Signed-off-by: Jess Frazelle <acidburn@microsoft.com>

jessfraz committed:
ProcMount: add dockershim support

Signed-off-by: Jess Frazelle <acidburn@microsoft.com>

feiskyer commented:
/status approved-for-milestone

dims commented:
/retest

redbaron commented:
/retest

jessfraz commented:
@smarterclayton or someone? can I get review?

jessfraz commented:
/test pull-kubernetes-verify

dims commented:
@jessfraz i don't think the verify error is caused by your PR it's from a pr that merged earlier today - https://github.com/kubernetes/kubernetes/commit/7b8651d9122aadcb5b31b6fabc40aa76befefdba - someone forgot to remove the dependencies

jimangel commented:
@jessfraz I'm a rookie and part of the 1.12 docs release team, I'm not sure I 100% understand the implementation for this, but it looks like it will need some docs (add'l api options / feature gates). Do you agree? If so, let me know if there's a PR for the docs so I can mark it down thanks!

jessfraz commented:
@jimangel  will do!

jessfraz commented:
@dims thanks!

dashpole commented:
cadvisor godep update lgtm

dims commented:
@jessfraz https://github.com/kubernetes/kubernetes/pull/66410 has merged, if we rebase this now, we should be all set with the verify job.

dims commented:
@yujuhong @dchen1107 fyi, one of the side results of the docker/docker update in this PR is that we switch from docker api version `1.31` to `1.38`

https://github.com/kubernetes/kubernetes/pull/64283/files#diff-a7d26e41cd477718386091cba0af8a79L18

dims commented on pkg/features/kube_features.go:
```diff
@@ -343,6 +343,12 @@ const (
 	CSIBlockVolume utilfeature.Feature = "CSIBlockVolume"
+
+	// owner: @jessfraz
+	// alpha: v1.11
```
v1.12

dims commented on pkg/securitycontext/util.go:
```diff
@@ -167,3 +172,48 @@ func AddNoNewPrivileges(sc *v1.SecurityContext) bool {
+
+var (
+	defaultMaskedPaths = []string{
+		"/proc/kcore",
```
Jess, looks like we could add `/proc/acpi` (from https://github.com/moby/moby/blob/86a41e4ef70ddeca610f2d0ed195a7834fc6bf4d/oci/defaults.go#L116-L134) may be even add a comment to keep these 2 lists in sync with the code in oci/defaults.go ?

dims reviewed (approved):
a femto-nit and a question inline.

tpepper reviewed (changes requested):
Would it be possible to get a longer commit message on the vendor code update commit describing what functionality/fixes were sought in which version of the dep versus what prior version?  We've had a few cases lately of a regression causing somebody else to spot via bisection and revert an insufficiently justified vendor code update commit, which then of course broke the thing the newer vendor code was brought in for.  A bit more documentation on the commit could save a number of people a lot of time along such ripples.

jessfraz commented:
@tpepper added more to commit message thanks! @dims comments addressed :)

tpepper reviewed (approved):
My requested small changes were made so an
/lgtm
from me.

dims commented:
/assign @thockin
/assign @derekwaynecarr

dims commented:
/assign @smarterclayton

dims commented:
/test pull-kubernetes-e2e-kops-aws
/test pull-kubernetes-e2e-gce

jimangel commented:
@jessfraz do you have a PR already for the docs?

@dims I see this bumps the required docker API version, do we need to update other docs (any requirements, etc)?

dims commented:
@jimangel that's definitely the easy, straightforward approach, i was toying with another approach in case there is a concern - https://github.com/kubernetes/kubernetes/pull/65872

jessfraz commented:
Opened https://github.com/kubernetes/website/pull/9911 for docs... and I will rebase

jessfraz committed:
vendor: update docker cadvisor winterm

This vendor change was purely for the changes in docker to allow for
setting the Masked and Read-only paths.

See: moby/moby#36644

But because of the docker dep update it also needed cadvisor to be
updated and winterm due to changes in pkg/tlsconfig in docker

See: google/cadvisor#1967

Signed-off-by: Jess Frazelle <acidburn@microsoft.com>

jessfraz commented:
rebased :)

dims commented:
/lgtm

jessfraz commented:
/test pull-kubernetes-e2e-gce

jessfraz commented:
woohoo its green \o/

jessfraz commented:
rebased, ping @smarterclayton ?

nikhita commented:
/remove-area custom-resources

thockin reviewed (commented):
Without having an opinion on the PR, the API seems fine.

Can you explain the tangle of godep changes?  Why are we updating ansiterm?

jessfraz commented:
@thockin  yeah I tried to explain in the commit message as well but to upgrade docker/docker we needed to upgrade cadvisor because it depended on tlsconfig in docker/docker which no longer existed, and the same was for the windows ansiterm package. it was a weird rabbit hole

jessfraz commented:
/test pull-kubernetes-verify

dims commented:
/test pull-kubernetes-e2e-kops-aws
/lgtm

nikhita commented:
Looks like this PR would be a good candidate for the `queue/multiple-rebases` label. :)

nikhita commented:
/remove-area custom-resources

jessfraz commented:
woohoo its green and rebased

dims commented:
/lgtm

jessfraz commented:
thanks @dims for bearing with me :)

jessfraz commented:
/assign @smarterclayton

smarterclayton commented on pkg/apis/core/types.go:
```diff
@@ -4594,8 +4594,27 @@ type SecurityContext struct {
+	DefaultProcMount ProcMountType = "Default"
+
+	// UnmaskedProcMount bypasses the default masking behavior of the container
+	// runtime and ensures the newly created /proc the container stays in tact with
```
nit: `intact`

smarterclayton commented on staging/src/k8s.io/api/core/v1/types.go:
```diff
@@ -5175,8 +5175,27 @@ type SecurityContext struct {
 	AllowPrivilegeEscalation *bool `json:"allowPrivilegeEscalation,omitempty" protobuf:"varint,7,opt,name=allowPrivilegeEscalation"`
+	// ProcMount denotes the type of proc mount to use for the containers.
+	// The default is DefaultProcMount which uses the container runtime defaults for
+	// readonly paths and masked paths.
```
Comment that this requires the flag gate to be enabled.

smarterclayton commented on staging/src/k8s.io/api/core/v1/types.go:
```diff
@@ -5175,8 +5175,27 @@ type SecurityContext struct {
 	// 2) has CAP_SYS_ADMIN
 	// +optional
 	AllowPrivilegeEscalation *bool `json:"allowPrivilegeEscalation,omitempty" protobuf:"varint,7,opt,name=allowPrivilegeEscalation"`
+	// ProcMount denotes the type of proc mount to use for the containers.
```
This should be lower cased when referenced in external API godoc (since that shows up in openapi and is displayed in tools where users only see the JSON field).

jessfraz replied:
the other ones above this were not lowercased so just wondering what you mean exactly :)

smarterclayton replied:
Basically the original API we didn't enforce it and we started doing it.  If you look at newer APIs we enforce that consistently.

jessfraz replied:
gotcha will update

smarterclayton commented on staging/src/k8s.io/api/policy/v1beta1/types.go:
```diff
@@ -221,6 +221,10 @@ type PodSecurityPolicySpec struct {
 	// +optional
 	ForbiddenSysctls []string `json:"forbiddenSysctls,omitempty" protobuf:"bytes,20,rep,name=forbiddenSysctls"`
+	// AllowedProcMountTypes is a whitelist of allowed ProcMountTypes.
+	// Empty or nil indicates that only the DefaultProcMountType may be used.
```
Is this also covered by the flag gate?  If so same comment as above.

smarterclayton commented on staging/src/k8s.io/api/extensions/v1beta1/types.go:
```diff
@@ -965,6 +965,10 @@ type PodSecurityPolicySpec struct {
+	// AllowedProcMountTypes is a whitelist of allowed ProcMountTypes.
+	// Empty or nil indicates that only the DefaultProcMountType may be used.
+	// +optional
+	AllowedProcMountTypes []v1.ProcMountType `json:"allowedProcMountTypes,omitempty" protobuf:"bytes,21,opt,name=allowedProcMountTypes"`
```
Same comment about flag gating.

smarterclayton commented on Godeps/Godeps.json:
```diff
@@ -1310,153 +1310,133 @@
-			"Comment": "docs-v1.12.0-rc4-2016-07-15-7401-g4f3616fb1",
-			"Rev": "4f3616fb1c112e206b88cb7a9922bf49067a7756"
+			"Comment": "docs-v1.12.0-rc4-2016-07-15-9510-ga9fbbdc8d",
+			"Rev": "a9fbbdc8dd8794b20af358382ab780559bca589d"
```
What effective docker version are we bumping to?  Has sig-node weighed in on this update yet?

jessfraz replied:
its 1.38.0 and see the comment below :) thanks @Random-Liu

smarterclayton reviewed (changes requested):
Some API doc changes. Want to make sure sig-node is good on the docker version bump, especially this close to release.

dashpole commented:
cc @Random-Liu for the vendored docker version bump.

Random-Liu commented:
The api packages used by `kube_docker_client.go`:
* github.com/docker/docker/api/types
* github.com/docker/docker/api/types/container
* github.com/docker/docker/api/types/image

The diff looks safe to me: https://gist.github.com/Random-Liu/629bb4a7be32de3208f2d99e2391ce27
So the logic that kubelet talks with docker shouldn't be affected.

However, if we reuse some docker package inside kubernetes, it is hard for me to evaluate the impact.

jessfraz committed:
address comments

Signed-off-by: Jess Frazelle <acidburn@microsoft.com>

jessfraz committed:
make update

Signed-off-by: Jess Frazelle <acidburn@microsoft.com>

dims commented:
/lgtm

smarterclayton commented on pkg/api/pod/util.go:
```diff
@@ -284,6 +286,24 @@ func DropDisabledRunAsGroupField(podSpec *api.PodSpec) {
 
+// DropDisabledProcMountField removes disabled fields from PodSpec related
+// to ProcMount
+func DropDisabledProcMountField(podSpec *api.PodSpec) {
```
I don't see the equivalent for DropDisabledAllowedProcMountTypes on PSP - am I just blind?

jessfraz replied:
fixed :)

jessfraz commented:
updated :)

feiskyer reviewed (approved):
LGTM

dims commented:
/lgtm

smarterclayton commented:
/approve

jessfraz commented:
/test pull-kubernetes-e2e-kops-aws

jessfraz commented:
/test pull-kubernetes-e2e-kops-aws

derekwaynecarr commented:
Can someone confirm this does not break support for docker 1.13?

annismckenzie commented:
It doesn't – as long as your pod specs don't request/use the feature. According to https://github.com/kubernetes/kubernetes/pull/64283/files#diff-9a9f71a82ddbbcd0627de3aea90c6649R262 it'll be available starting with Docker API version 1.38 (which isn't even in a released version yet – see [here](https://docs.docker.com/engine/api/version-history/)). At that point in the code you also see that it'll fail to admit pods that ask for the feature with Docker not being at that version.

dims commented:
@derekwaynecarr as i mentioned earlier in this PR - `one of the side results of the docker/docker update in this PR is that we switch from docker api version 1.31 to 1.38`  if we want to keep the max to the earlier 1.31 docker API version then we need something like:
https://github.com/kubernetes/kubernetes/pull/66511/

yujuhong commented:
I know the PR is already in, but I'd rather not add more Docker API checking logic in kubelet itself (with dockershim being the exception). Can we move that logic out of kubelet, and just let dockershim return an error when trying to start the pod? The pod will be pending but it's no worse than what this PR currently does IMO.
